types/types.go

```diff
@@ -218,6 +225,23 @@ type BlobInfoCache interface {
 	CandidateLocations(transport ImageTransport, scope BICTransportScope, digest digest.Digest, canSubstitute bool) []BICReplacementCandidate
 }
 
+// BlobInfoCache2 extends BlobInfoCache by adding the ability to track information about what kind of compression was applied to
+// the blobs it keeps information about.  If a BlobInfoCache also implements BlobInfoCache2, users of the cache are encouraged to
+// use RecordKnownLocation2() instead of RecordKnownLocation(), and to use CandidateLocations2() instead of CandidateLocations().
+type BlobInfoCache2 interface {
```

Comment:
My initial impression is that this shouldn’t be public; we might be forced to open it up again, but let’s at least try not to.

- Move both `BlobInfoCache2` and `BICReplacementCandidate2` into `c/image/internal/blobinfocache`
  - (Be extra paranoid and add a non-public method that is impossible to implement outside of c/image/internal ? That might be overdoing it to the point of being actively unhelpful.)
- Document `BlobInfoCache` as (supported but?) deprecated in the sense that library callers should use one of the cache implementations in `c/image/pkg/blobinfocache` only, if at all possible.
- In `c/image/internal/blobinfocache`, provide helper functions that abstract the difference for callers, so that every transport doesn’t have to open-code that.
  - Non-blocking: It could also be possible to abstract the difference for the implementations (i.e. provide a “forwarding” implementation of `RecordKnownLocation` that can be embedded in a cache implementation).
  - (That _might_ be also a place to implement more of the common logic, perhaps even “update `BlobInfo` from a stream” and the like. I’m not at all sure yet, naively without taking a deeper look I’d prefer most of the intelligence to be not the responsibility of the transports but of `c/image/copy`.)

Reply:
Sounds good to me.  Will make those changes.

Reply:
Okay, removed a lot of the "if implements v2, use it, else call the v1 version" logic in favor of wrapping cache implementations that only implement one or the other pair of methods in a cache object that provides both sets of methods, and using that everywhere.